packages/router/src/lib/operations/auction.ts

```diff
@@ -185,8 +173,75 @@ export const newAuction = async (
     });
   }
 
+  let sendingChainIdx: number = -1;
+  let receivingChainIdx: number = -1;
+  let swapPoolIndex: number = -1;
+  const allowedSwap = config.swapPools.find((pool, index) => {
+    const existSwap =
+      pool.assets.find((a) => getAddress(a.assetId) === getAddress(sendingAssetId) && a.chainId === sendingChainId) &&
+      pool.assets.find((a) => getAddress(a.assetId) === getAddress(receivingAssetId) && a.chainId === receivingChainId);
+
+    if (existSwap) {
+      sendingChainIdx = pool.assets.findIndex(
+        (a) => getAddress(a.assetId) === getAddress(sendingAssetId) && a.chainId === sendingChainId,
+      );
+
+      receivingChainIdx = pool.assets.findIndex(
+        (a) => getAddress(a.assetId) === getAddress(receivingAssetId) && a.chainId === receivingChainId,
+      );
+      swapPoolIndex = index;
+    }
+
+    return existSwap;
+  });
+  if (!allowedSwap) {
+    throw new SwapInvalid(sendingChainId, sendingAssetId, receivingChainId, receivingAssetId, {
+      methodContext,
+      requestContext,
+    });
+  }
+
+  const swapPool = config.swapPools[swapPoolIndex];
+  const routerBalances = await Promise.all(
+    swapPool.assets.map(async (asset) => {
+      const assetLiquidity = await contractReader.getAssetBalance(asset.assetId, asset.chainId);
+      let assetDecimals = chainData.get(asset.chainId.toString())?.assetId[asset.assetId]?.decimals;
+      if (!assetDecimals) {
+        assetDecimals = await txService.getDecimalsForAsset(asset.chainId, asset.assetId);
+      }
+      // convert asset liquidity into 18 decimal value.
+      const res = assetLiquidity.mul(BigNumber.from(10).pow(18 - assetDecimals));
+      return res;
+    }),
+  );
+
+  logger.info("Got balances of router", requestContext, methodContext, {
+    routerBalances: routerBalances.map((balance) => balance.toString()),
+    sendingChainIdx,
+    receivingChainIdx,
+  });
+
+  const [senderLiquidity, receiverLiquidity] = await Promise.all([
+    contractReader.getAssetBalance(sendingAssetId, sendingChainId),
+    contractReader.getAssetBalance(receivingAssetId, receivingChainId),
+  ]);
+
+  logger.info("Got asset liquidity", requestContext, methodContext, {
+    senderLiquidity: senderLiquidity.toString(),
+    receiverLiquidity: receiverLiquidity.toString(),
+  });
+
   // getting the swap rate from the receiver side config
-  let amountReceived = await getReceiverAmount(amount, inputDecimals, outputDecimals);
+
+  let amountReceived = await getReceiverAmount(
+    amount,
+    inputDecimals,
+    outputDecimals,
+    routerBalances,
```

Comment:
we should probably have the conversion to 18 decimals from whatever it has onchain happen in only one place, so in the `getReceiverAmount` function only

Reply:
sorry?

Reply:
wow highlighted the incorrect bit of code 😅 

Same as I mentioned above -- there's only one function that requires decimal manipulation, so lets make sure we keep the manipulation isolated to that function